### PR TITLE
Adds support for checking the timestamp inherent while validating a block

### PR DESCRIPTION
**Pull request by bkchr**
This adds support for checking the timestamp inherent while validating a
block. This will use the relay chain slot number * relay chain slot
duration to calculate a timestamp. This timestamp is used to check the
timestamp in the timestamp inherent.

Fixes: https://github.com/paritytech/cumulus/issues/422

### CODE REVIEW

File: polkadot-parachains/rococo-runtime/src/lib.rs [hunk 1]
```diff
@@ -595,10 +592,22 @@ struct CheckInherents;
...
 	fn check_inherents(
-		_: &[UncheckedExtrinsic],
-		_: &cumulus_pallet_parachain_system::RelayChainStateProof,
+		block: &Block,
+		relay_state_proof: &cumulus_pallet_parachain_system::RelayChainStateProof,
 	) -> sp_inherents::CheckInherentsResult {
-		sp_inherents::CheckInherentsResult::new()
+		let relay_chain_slot = relay_state_proof
+			.read_slot()
+			.expect("Could not read the relay chain slot from the proof");
```
Comment (JoshOrndorff) [inline polkadot-parachains/rococo-runtime/src/lib.rs:600]: This check depends on the relay state proof which came in through the parachain inherent.

Maybe a dumb question, but do the validators check that the data in the parachain inherent reflects the real relay chain state?

Comment (bkchr) [reply polkadot-parachains/rococo-runtime/src/lib.rs:600]: Yes

File: polkadot-parachains/rococo-runtime/src/lib.rs [hunk 2]
```diff
@@ -340,18 +339,16 @@ impl Config for XcmConfig {
...
 	type Trader = UsingComponents<IdentityFee<Balance>, RocLocation, AccountId, Balances, ()>;
-	type ResponseHandler = ();	// Don't handle responses for now.
+	type ResponseHandler = (); // Don't handle responses for now.
 }
 
 /// No local origins on this chain are allowed to dispatch XCM sends/executions.
-pub type LocalOriginToLocation = (
-	SignedToAccountId32<Origin, AccountId, RococoNetwork>,
-);
+pub type LocalOriginToLocation = (SignedToAccountId32<Origin, AccountId, RococoNetwork>,);
```
Comment (shawntabrizi) [inline polkadot-parachains/rococo-runtime/src/lib.rs:351]:
```suggestion
pub type LocalOriginToLocation = (SignedToAccountId32<Origin, AccountId, RococoNetwork>);
```

File: polkadot-parachains/rococo-runtime/src/lib.rs [hunk 3]
```diff
@@ -595,10 +592,22 @@ struct CheckInherents;
...
 	) -> sp_inherents::CheckInherentsResult {
-		sp_inherents::CheckInherentsResult::new()
+		let relay_chain_slot = relay_state_proof
+			.read_slot()
+			.expect("Could not read the relay chain slot from the proof");
+
+		let inherent_data =
+			cumulus_primitives_timestamp::InherentDataProvider::from_relay_chain_slot_and_duration(
+				relay_chain_slot,
+				sp_std::time::Duration::from_secs(6),
```
Comment (shawntabrizi) [inline polkadot-parachains/rococo-runtime/src/lib.rs:605]: this needs to be hardcoded?

File: polkadot-parachains/statemine-runtime/src/lib.rs [hunk 4]
```diff
@@ -149,9 +147,10 @@ parameter_types! {
...
 impl Filter<Call> for BaseFilter {
 	fn filter(c: &Call) -> bool {
-		!matches!(c,
-			Call::Assets(pallet_assets::Call::create(..)) |
-			Call::Uniques(pallet_uniques::Call::create(..))
+		!matches!(
+			c,
+			Call::Assets(pallet_assets::Call::create(..))
+				| Call::Uniques(pallet_uniques::Call::create(..))
 		)
```
Comment (shawntabrizi) [inline polkadot-parachains/statemine-runtime/src/lib.rs:154]: this kind of formatting is why we will never get rustfmt

File: polkadot-parachains/statemine-runtime/src/lib.rs [hunk 5]
```diff
@@ -557,22 +561,20 @@ impl Config for XcmConfig {
...
 
 parameter_types! {
 	pub const MaxDownwardMessageWeight: Weight = MAXIMUM_BLOCK_WEIGHT / 10;
 }
 
 /// No local origins on this chain are allowed to dispatch XCM sends/executions.
-pub type LocalOriginToLocation = (
-	SignedToAccountId32<Origin, AccountId, RelayNetwork>,
-);
+pub type LocalOriginToLocation = (SignedToAccountId32<Origin, AccountId, RelayNetwork>,);
```
Comment (shawntabrizi) [inline polkadot-parachains/statemine-runtime/src/lib.rs:577]:
```suggestion
pub type LocalOriginToLocation = (SignedToAccountId32<Origin, AccountId, RelayNetwork>);
```

File: polkadot-parachains/statemint-runtime/src/lib.rs [hunk 6]
```diff
@@ -831,10 +830,22 @@ struct CheckInherents;
...
 	) -> sp_inherents::CheckInherentsResult {
-		sp_inherents::CheckInherentsResult::new()
+		let relay_chain_slot = relay_state_proof
+			.read_slot()
+			.expect("Could not read the relay chain slot from the proof");
+
+		let inherent_data =
+			cumulus_primitives_timestamp::InherentDataProvider::from_relay_chain_slot_and_duration(
+				relay_chain_slot,
+				sp_std::time::Duration::from_secs(6),
```
Comment (shawntabrizi) [inline polkadot-parachains/statemint-runtime/src/lib.rs:843]: also must be hardcoded here?

File: polkadot-parachains/rococo-runtime/src/lib.rs (hunk 3 above)
Comment (andresilva) [reply polkadot-parachains/rococo-runtime/src/lib.rs:605]: Maybe create a const for `RELAY_SLOT_DURATION`.

File: primitives/timestamp/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	/// Create the inherent data.
+	pub fn create_inherent_data(&self) -> Result<InherentData, Error> {
+		let mut inherent_data = InherentData::new();
+		self.provide_inherent_data(&mut inherent_data)
+			.map(|_| inherent_data)
+	}
+
+	/// Provide the inherent data into the given `inherent_data`.
+	pub fn provide_inherent_data(&self, inherent_data: &mut InherentData) -> Result<(), Error> {
+		// As the parachain starts to building at around `relay_chain_slot + 1` we use that slot to
```
Comment (andresilva) [inline primitives/timestamp/src/lib.rs:66]:
```suggestion
		// As the parachain starts building at around `relay_chain_slot + 1` we use that slot to
```

**Review by andresilva — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**